### PR TITLE
Add dataType option to Select fields

### LINKED ISSUE #398: Should `Select` field options be a GraphQL `String` or `Enum`?
In GraphQL, `Enum` values are restricted to those that can be described by the regex `/^[_A-Za-z][_0-9A-Za-z]*$/`. That is, they must start with an underscore or a letter and can contain only letters and digits.

I _think_ we want `Select` options to be less restrictive that this, no?

If not, we should be doing some fairly tight validation of the field config options and returning sensible errors.

### LINKED ISSUE #851: Select option expansion logic for arrays of strings is broken
I am creating a `Player` list with a `position` field using a `Select` type.

In my select, I have the 5 positions on a basketball court:

```javascript
keystone.createList("Player", {
  fields: {
    position: {
      type: Select,
      options: [
        "Point Guard",
        "Shooting Guard",
        "Small Forward",
        "Power Forward",
        "Center"
      ]
    }
  }
});
```

Doing so triggers the following errors:

```
Error: 
Enum value "PlayerPositionType.Guard" can only be defined once.
Enum value "PlayerPositionType.Forward" can only be defined once.
```

Same deal happens if I pass the options as an array of objects, or a comma separated string.

FWI My workaround was to pass a comma separated string with underscores instead of spaces (`"point_guard, shooting_guard, small_forward, power_forward, center"`) and let the `humanize()` method format it, but I don't think it's ideal!

### LINKED ISSUE #1768: A select with number at start of value throws an error
# Bug report

## Describe the bug

If you create a list with numbers at the start of the value of will throw an error on start.

` message: 'Syntax Error: Expected Name, found Int "0"',`

## To Reproduce

1. Create a list like: 

```
keystone.createList('ExampleList', {
  fields: {
    age: {
      type: Select,
      options: '0-18,19-35,36-50,50+',
    },
  },
});
```
2. Start Keystone.

It doesn't seem to matter if you use the array notation either:
```
keystone.createList('ExampleList', {
  fields: {
    age: {
      type: Select,
      options: [
        { label: '0-18', value: '0-18' },
        { label: '19-35', value: '19-35' },
        { label: '36-50', value: '36-50' },
        { label: '50+', value: '50+' },
      ],
    },
  },
});
```

## Expected behaviour

I can't see a clear reason why this cannot work. There are no limitations on storing this in a text field in the DB or on the possible values in HTML.

## System information

I was using Knex Adapter

## Additional context

Is Monday.

### PR DESCRIPTION
**Pull request by JedWatson**
Been meaning to implement this for a long time.

Our Select field should support storing value as an `enum` / `string` / `integer` with the corresponding GraphQL data type. Now it does!

Fixes #851
Fixes #398 
Fixes #1768

### CODE REVIEW

**Comment by MadeByMike** [comment]
I did a smoke test and admin UI looks ok. I assume this might break if the AdminUI defined a graphQL query that was expecting an enum, and it looks like it doesn't hard code this.

**Comment by JedWatson** [comment]
@MadeByMike I think you're right - I have tested the Admin UI but will review the controller again for how filters are submitted.

**Comment by MadeByMike** [comment]
This might be overkill but I've added filter tests for each of the data types. All passing.

File: packages/fields/src/types/Select/test-fixtures.js [hunk 1]
```diff
@@ -53,13 +71,77 @@ export const filterTests = withKeystone => {
...
+          { name: 'c', selectNumber: 3 },
+          { name: 'd', selectNumber: 1 },
+        ],
+        'name selectNumber'
+      )
+    )
+  );
+
   test(
     'Filter: company',
```
Comment (timleslie) [inline packages/fields/src/types/Select/test-fixtures.js:109]: Can we add `(dataType: enum)` on all of the original tests just for consistency?

**Comment by timleslie** [comment]
Just from eyeballing the code, it looks to me like `getFilterGraphQL` won't do the right thing when using `dataType = string`. My reading of it is that we should be putting double quotes around `value` in that case.

**Comment by JedWatson** [comment]
Fixed the filtering in the Admin UI, added some more helpful type validation for options, and added explicit `enum` case for the tests.

I think this is good to merge now

**Pull request merged**